### DISCUSSION TITLE
Figure cross referencing and  knitr::read_chunk

### DISCUSSION THREAD

**Discussion by richardbeare** [quarto-dev/quarto-cli#8758] [Q&A] [answered]
### Description

Hi,
I'm a fan of having my R code in chunks in a separate file and including what I need in the Rmd file as required. However I'm struggling to get figure labels to work when I do this with quarto. I can't see what I'm missing here - no matter what I try I can't get any version of ```@fig-label``` to work. The chunks import just fine - the label/cross reference is the missing part. Any suggestions?

````qmd

---
title: "Test figure labels from code chunks"
format:
  pdf:
    keep-md: true
    keep-tex: true
---

## Quarto
```{r}
#| echo: FALSE
knitr::read_chunk("qq_code.R")
```

```{r }
#| label: Setup
#| echo: FALSE
#| warning: false
#| message: false
```

```{r}
#| label: APlot
#| echo: FALSE
#| fig.cap: "A caption"
#| fig.link: fig-APlot
```

Figure @APlot

````

and `qq_code.R` :

```r
## ---- Setup ----
library(tidyverse)

data(mtcars)

## ---- APlot ----
ggplot(mtcars, aes(x=disp, y=mpg, colour=factor(cyl))) + geom_point()

```

**Comment by mcanouil** [comment, marked as answer]
Label needs a prefix such as `fig-` to be accounted as a cross-reference, see <https://quarto.org/docs/authoring/cross-references.html>.

Side note: `FALSE` is R syntax. In yaml, it is `false`. In Quarto, options use `-` not `.` as word separator.
Use `execute` to set globally option such as `echo: false`, see <https://quarto.org/docs/reference/formats/html.html#execution> and <https://quarto.org/docs/computations/execution-options.html>.
Finally, headers, list, and code cells should be surrounded by empty lines to avoid parsing issues.

In the end your Quarto document should look like the following:

````qmd
---
title: "Test figure labels from code chunks"
format:
  pdf:
    keep-md: true
    keep-tex: true
execute:
  echo: false
---

## Quarto

```{r}
#| include: false
knitr::read_chunk("qq_code.R")
```

```{r}
#| label: Setup
#| include: false
#| warning: false
#| message: false
```

```{r}
#| label: fig-APlot
#| ref-label: APlot
#| fig-cap: "A caption"
#| fig-link: fig-APlot
```

Figure @fig-APlot
````

**Reply by richardbeare** [reply]
Thanks, that seems to work, and means that chunks in the R code producing figures need to have a specific naming structure. i.e. fig-something in the R file. Didn't realise that.

**Reply by mcanouil** [reply]
I updated to fit your R script, using `ref-label` knitr option.

**Reply by richardbeare** [reply]
Hm, I can't get that approach to work. It is OK if I rename the R chunks to include a fig- prefix, but not via ref-label

**Reply by mcanouil** [reply]
What do you mean?
Which version of `knitr` and Quarto CLI are you using?

The approach works.
<img width="1624" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/369efd38-f5b7-45f5-94b0-d55a74dc86de">

**Reply by richardbeare** [reply]
Must be a version problem - older on my linux machine where I was doing the work - Quarto 1.3.353, knitr 1.43, rmarkdown 2.23, pandoc 3.1.1. Seems OK on my mac - Quarto 1.3.433, knitr 1.44. Time for upgrades all round.

Thanks
